[OpenXR] Use AHardwareBuffer for buffer sharing on Android

aperezdc opened the pull request:
#### f59489c669a20aededb0e299e54d3ed448424b9b
<pre>
[OpenXR] Use AHardwareBuffer for buffer sharing on Android
<a href="https://bugs.webkit.org/show_bug.cgi?id=298312">https://bugs.webkit.org/show_bug.cgi?id=298312</a>

Reviewed by Carlos Garcia Campos.

Add a new GraphicsContextGLTextureMapperAndroid class, which mostly
reuses the functionality from the GraphicsContextGLTextureMapperANGLE;
but adds its own implementation of methods used by the WebXR machinery
to use GCGLExternalImage instances backed by AHardwareBuffer that can be
shared with the rendering process (currently WebProcess, but could be
GPUProcess). Buffer sharing across processes uses the support added
in 299311@main.

It is not possible to extract the AHardwareBuffer backing an existing
texture provided by the OpenXR runtime as part of its swapchain (and
textures may not even use one), so instead new compatible buffers are
created where the actual rendering is done, and then the result blitted
onto the OpenXR-provided textures. This reuses the existing blitting
mechanism used as fallback when the MESA_image_dma_buf_export is not
available, which was introduced in introduced in 299044@main.

The last bit of the piece is handling XrInstanceCreateInfoAndroidKHR
during OpenXR initialization: the runtime needs to know which JavaVM
instance for the application, and the Activity from the Android UI
framework to associate with the XrInstance. To obtain these, the code
expects the following two functions to be defined and resolvable at
run time via dlsym(RTLD_DEFAULT, ...):

  JavaVM* wpe_android_runtime_get_current_java_vm()
  jobject wpe_android_runtime_get_current_activity()

The WPE-Android project has incorporated those in the following patch:

  <a href="https://github.com/Igalia/wpe-android/pull/214">https://github.com/Igalia/wpe-android/pull/214</a>

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::makeExternalImageSource):
(WebCore::WebXROpaqueFramebuffer::reusableDisplayAttachments const):
(WebCore::WebXROpaqueFramebuffer::bindCompositorTexturesForDisplay):
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/android/GraphicsContextGLTextureMapperAndroid.cpp: Added.
(WebCore::GraphicsContextGLTextureMapperAndroid::create):
(WebCore::GraphicsContextGLTextureMapperAndroid::platformInitializeExtensions):
(WebCore::GraphicsContextGLTextureMapperAndroid::createExternalImage):
(WebCore::GraphicsContextGLTextureMapperAndroid::bindExternalImage):
(WebCore::GraphicsContextGLTextureMapperAndroid::enableRequiredWebXRExtensions):
* Source/WebCore/platform/graphics/android/GraphicsContextGLTextureMapperAndroid.h: Added.
* Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
(WebCore::GLDisplay::GLDisplay):
* Source/WebCore/platform/graphics/egl/GLDisplay.h:
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::createWebProcessGraphicsContextGL):
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
(WebKit::OpenXRLayer::exportOpenXRTextureAndroid):
(WebKit::OpenXRLayer::exportOpenXRTexture):
(WebKit::OpenXRLayerProjection::endFrame):
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.h:
(WebKit::OpenXRLayer::needsBlitTexture const):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::createInstance):
(WebKit::OpenXRCoordinator::createGLDisplay const):
(WebKit::OpenXRCoordinator::tryInitializeGraphicsBinding):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: <a href="https://commits.webkit.org/300466@main">https://commits.webkit.org/300466@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9bbb2fb77663cac163c8358e657115d42c49e9f9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122652 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42364 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33054 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129270 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74756 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124528 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50957 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93223 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125604 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34345 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109803 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73867 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33332 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27959 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72756 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104042 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28170 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131993 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49597 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37747 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/101756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49973 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106019 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/101621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25805 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46986 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25154 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46367 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49454 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55207 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48921 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52273 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50604 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ae9abcde9c3082faf52f85ed34f41037a5c40185. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50857

aperezdc commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -63,6 +69,88 @@ OpenXRLayer::~OpenXRLayer()
+        return { };
+    }
+
+    GLint boundTexture = 0;
+    glGetIntegerv(GL_TEXTURE_BINDING_2D, &boundTexture);
+    PlatformGLObject exportedTexture;
+    glGenTextures(1, &exportedTexture);
+    glBindTexture(GL_TEXTURE_2D, exportedTexture);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, image);
+    glBindTexture(GL_TEXTURE_2D, boundTexture);
+
+    display.destroyImage(image);
```
We can always do the refactor in a follow-up patch if we want to clean up this a bit.

spenap commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -353,12 +359,29 @@ void OpenXRCoordinator::createInstance()
+    auto java = createOpenXRStruct<XrInstanceCreateInfoAndroidKHR, XR_TYPE_INSTANCE_CREATE_INFO_ANDROID_KHR>();
+    java.applicationVM = s_wpeAndroidRuntimeGetJavaVM();
+    java.applicationActivity = s_wpeAndroidRuntimeGetActivity();
+    createInfo.next = (XrBaseInStructure*)&java;
```
could this use a `reinterpret_cast`?

carlosgcampos commented on Source/WebCore/platform/graphics/android/GraphicsContextGLTextureMapperAndroid.h:
```diff
@@ -0,0 +1,53 @@
+#include "GraphicsContextGLTextureMapperANGLE.h"
+#include <wtf/android/RefPtrAndroid.h>
+
+extern "C" {
+    typedef void* EGLImageKHR;
+};
```
Why is this needed?

carlosgcampos commented on Source/WebCore/platform/xr/PlatformXR.h:
```diff
@@ -310,8 +317,11 @@ struct FrameData {
+        explicit operator bool() const { return !fds.isEmpty(); }
 #endif
     };
+#endif // OS(ANDROID)
```
I find this a bit confusing when there's an else. I think I would omit the comment.

carlosgcampos commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -8942,11 +8942,15 @@ using WebCore::GraphicsContextGL::ExternalSyncSource = std::tuple<MachSendRight,
     Vector<uint32_t> offsets;
     uint32_t fourcc;
     uint64_t modifier;
+#endif // OS(ANDROID)
```
Same here, and this is a smaller block.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.h:
```diff
@@ -23,6 +23,10 @@
 
 #include "OpenXRUtils.h"
 
+#if OS(ANDROID)
+#define jobject void*
+#endif
```
What is this jobject?

aperezdc replied:
It's the type that the Java Native Interface (JNI) uses for handles to Java objects managed by the JVM.

aperezdc replied:
On second thoughts, on Android we can rely on `<jni.h>` always being available in the NDK, so I suppose it's not a big deal to include the header, guarder by `OS(ANDROID)`.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -121,6 +212,7 @@ std::optional<PlatformXR::FrameData::ExternalTexture> OpenXRLayer::exportOpenXRT
         .modifier = modifier,
     };
 }
+#endif // OS(ANDROID)
```
Same about the comment, either omit or use `!OS(ANDROID)` because this is the end of the non android block.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -337,7 +348,7 @@ void OpenXRCoordinator::createInstance()
     ASSERT(RunLoop::isMain());
     ASSERT(m_instance == XR_NULL_HANDLE);
 
-    Vector<char *, 2> extensions;
+    Vector<char *> extensions;
```
Why removing the inline size?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9bbb2fb77663cac163c8358e657115d42c49e9f9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50857

aperezdc committed:
[OpenXR] Use AHardwareBuffer for buffer sharing on Android
https://bugs.webkit.org/show_bug.cgi?id=298312

Reviewed by Carlos Garcia Campos.

Add a new GraphicsContextGLTextureMapperAndroid class, which mostly
reuses the functionality from the GraphicsContextGLTextureMapperANGLE;
but adds its own implementation of methods used by the WebXR machinery
to use GCGLExternalImage instances backed by AHardwareBuffer that can be
shared with the rendering process (currently WebProcess, but could be
GPUProcess). Buffer sharing across processes uses the support added
in 299311@main.

It is not possible to extract the AHardwareBuffer backing an existing
texture provided by the OpenXR runtime as part of its swapchain (and
textures may not even use one), so instead new compatible buffers are
created where the actual rendering is done, and then the result blitted
onto the OpenXR-provided textures. This reuses the existing blitting
mechanism used as fallback when the MESA_image_dma_buf_export is not
available, which was introduced in introduced in 299044@main.

The last bit of the piece is handling XrInstanceCreateInfoAndroidKHR
during OpenXR initialization: the runtime needs to know which JavaVM
instance for the application, and the Activity from the Android UI
framework to associate with the XrInstance. To obtain these, the code
expects the following two functions to be defined and resolvable at
run time via dlsym(RTLD_DEFAULT, ...):

  JavaVM* wpe_android_runtime_get_current_java_vm()
  jobject wpe_android_runtime_get_current_activity()

The WPE-Android project has incorporated those in the following patch:

  https://github.com/Igalia/wpe-android/pull/214

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::makeExternalImageSource):
(WebCore::WebXROpaqueFramebuffer::reusableDisplayAttachments const):
(WebCore::WebXROpaqueFramebuffer::bindCompositorTexturesForDisplay):
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/android/GraphicsContextGLTextureMapperAndroid.cpp: Added.
(WebCore::GraphicsContextGLTextureMapperAndroid::create):
(WebCore::GraphicsContextGLTextureMapperAndroid::platformInitializeExtensions):
(WebCore::GraphicsContextGLTextureMapperAndroid::createExternalImage):
(WebCore::GraphicsContextGLTextureMapperAndroid::bindExternalImage):
(WebCore::GraphicsContextGLTextureMapperAndroid::enableRequiredWebXRExtensions):
* Source/WebCore/platform/graphics/android/GraphicsContextGLTextureMapperAndroid.h: Added.
* Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
(WebCore::GLDisplay::GLDisplay):
* Source/WebCore/platform/graphics/egl/GLDisplay.h:
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::createWebProcessGraphicsContextGL):
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
(WebKit::OpenXRLayer::exportOpenXRTextureAndroid):
(WebKit::OpenXRLayer::exportOpenXRTexture):
(WebKit::OpenXRLayerProjection::endFrame):
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.h:
(WebKit::OpenXRLayer::needsBlitTexture const):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::createInstance):
(WebKit::OpenXRCoordinator::createGLDisplay const):
(WebKit::OpenXRCoordinator::tryInitializeGraphicsBinding):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: https://commits.webkit.org/300466@main
